src/BlockTypePalette.cpp

```diff
@@ -196,89 +220,150 @@ void BlockTypePalette::loadFromJsonString(const AString & aJsonPalette)
 
 void BlockTypePalette::loadFromTsv(const AString & aTsvPalette, bool aIsUpgrade)
 {
-	auto lines = StringSplitAndTrim(aTsvPalette, "\n");
+	static const AString hdrTsvRegular = "BlockTypePalette";
+	static const AString hdrTsvUpgrade = "UpgradeBlockTypePalette";
+
+	// Check the file signature:
+	auto idx = findNextSeparator(aTsvPalette);
+	if ((idx == AString::npos) || (aTsvPalette[idx] == '\t'))
+	{
+		throw LoadFailedException("Invalid signature");
+	}
+	auto signature = aTsvPalette.substr(0, idx);
+	bool isUpgrade = (signature == hdrTsvUpgrade);
+	if (!isUpgrade && (signature != hdrTsvRegular))
+	{
+		throw LoadFailedException("Unknown signature");
+	}
+	if (aTsvPalette[idx] == '\r')   // CR of the CRLF pair, skip the LF:
+	{
+		idx += 1;
+	}
 
 	// Parse the header:
-	int fileVersion = 0;
+	bool hasHadVersion = false;
 	AString commonPrefix;
-	auto numLines = lines.size();
-	for (size_t idx = 1; idx < numLines; ++idx)
+	int line = 2;
+	auto len = aTsvPalette.length();
+	while (true)
 	{
-		const auto & line = lines[idx];
-		if (line.empty())
+		auto keyStart = idx;
+		auto keyEnd = findNextSeparator(aTsvPalette, idx + 1);
+		if (keyEnd == AString::npos)
+		{
+			throw LoadFailedException(Printf("Invalid header key format on line %u", line));
+		}
+		if (keyEnd == idx + 1)  // Empty line, end of headers
 		{
-			// End of headers, erase them from lines[] and go parse the data
-			lines.erase(lines.begin(), lines.begin() + static_cast<AStringVector::difference_type>(idx) + 1);
+			if (aTsvPalette[keyEnd] == '\r')  // CR of the CRLF pair, skip the LF:
+			{
+				++keyEnd;
+			}
+			idx = keyEnd;
+			++line;
 			break;
 		}
-		auto s = StringSplit(line, "\t");
-		if (s.size() != 2)
+		auto valueEnd = findNextSeparator(aTsvPalette, keyEnd + 1);
+		if ((valueEnd == AString::npos) || (aTsvPalette[valueEnd] == '\t'))
 		{
-			throw LoadFailedException(Printf("Invalid header format on line %u", idx + 1));
+			throw LoadFailedException(Printf("Invalid header value format on line %u", line));
 		}
-		if (s[0] == "FileVersion")
+		auto key = aTsvPalette.substr(keyStart + 1, keyEnd - keyStart - 1);
```

Comment:
Why is `keyStart+1` the actual start of the key?

Reply:
True, that's a bit of a misnomer, all the found indices are for the separator just before the field.